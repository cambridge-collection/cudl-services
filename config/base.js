var config = {}

//Application settings and paths
config.appDir = '/home/cudl/node/cudl-services2';
config.dataDir = '/mnt/cudl-data';
config.cacheDir = '/mnt/cache';

//MySQL settings for cudl database services 
config.mysqlPool = 10;
config.mysqlHost = 'found-dom01.lib.cam.ac.uk';
config.mysqlUser = 'viewerdevuser';
config.mysqlPass = 'resuvedreweiv';
config.mysqlData = 'viewerdev';

//Image server
config.imageServer = 'http://172.22.83.199/iipsrv/iipsrv.fcgi?iiif=';
module.exports = config;
