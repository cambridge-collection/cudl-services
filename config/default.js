var config = {};

//Application settings and paths
config.appDir = '/home/cms93/cudl-services';
config.dataDir = '/mnt/cudl-data';
config.cacheDir = '/mnt/cache';
config.dcpdataDir = '/mnt/dcp-data';
config.user = 'chris';
config.group = 'chris';

//Postgres settings for cudl database services
config.postHost = 'cudl-postgres.cmzjzpssbgnq.eu-west-1.rds.amazonaws.com';
config.postUser = 'cudldev';
config.postPass = 'anwqxJcONu1NjUl';
config.postDatabase = 'viewerdev';

config.users = {};
config.users["1a1c31e8-0c04-11e4-b6a8-bf55e15d8c9d"] = { username: 'cudl-dev', password: 'secret', email: 'cudl-admin@lib.cam.ac.uk' };
config.users["385784fa-0c04-11e4-b5cc-a3745fee5631"] = { username: 'cudl-live', password: 'secret', email: 'cudl-admin@lib.cam.ac.uk' };
config.users["436b3396-0c04-11e4-b7ed-0308c2efa56c"] = { username: 'genizah', password: 'secret', email: 'genizah@lib.cam.ac.uk' };
config.users["51ce1fb4-0d91-11e4-ad30-77b11da19985"] = { username: 'huw', password: 'secret', email: 'cudl-admin@lib.cam.ac.uk' };

//Image server
config.imageServer = 'http://172.22.83.199/iipsrv/iipsrv.fcgi?iiif=';
config.darwinXTF = 'http://dcp-dev.lib.cam.ac.uk';

module.exports = config;

// Tagging
config.defaultRemoveRatio = 1/5;
