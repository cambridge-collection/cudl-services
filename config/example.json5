{
  dataDir: "/mnt/cudl-data",
  darwinXTF: "https://dcp-public.lib.cam.ac.uk/",
  legacyDcpDataDir: "/mnt/dcp-data/letters",

  xtfBase: "https://cudl-dev.lib.cam.ac.uk/xtf/",
  xtfIndexPath: "/usr/local/indices/index-cudl",

  postHost: "localhost",
  postDatabase: "viewerdev",
  postUser: "cudlservices",
  postPass: "secret",

  users: {
    // The key in the users object is the API key for the user
    "96ecb691-bdea-4720-8ae0-045875006ef7": { username: 'example', password: 'secret', email: 'admin@example.com' }
  }
}
