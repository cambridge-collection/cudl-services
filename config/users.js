var users = {};

users["1a1c31e8-0c04-11e4-b6a8-bf55e15d8c9d"] = { username: 'cudl-dev', password: 'secret', email: 'cudl-admin@lib.cam.ac.uk' };
users["385784fa-0c04-11e4-b5cc-a3745fee5631"] = { username: 'cudl-live', password: 'secret', email: 'cudl-admin@lib.cam.ac.uk' };
users["436b3396-0c04-11e4-b7ed-0308c2efa56c"] = { username: 'genizah', password: 'secret', email: 'genizah@lib.cam.ac.uk' };
users["51ce1fb4-0d91-11e4-ad30-77b11da19985"] = { username: 'huw', password: 'secret', email: 'cudl-admin@lib.cam.ac.uk' };
module.exports = users;
