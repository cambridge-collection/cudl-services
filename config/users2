var users = [
    { id: 1, username: 'cudl-dev', password: 'secret', email: 'cudl-admin@lib.cam.ac.uk', apikey: '1a1c31e8-0c04-11e4-b6a8-bf55e15d8c9d' }
  , { id: 2, username: 'cudl-live', password: 'secret', email: 'cudl-admin@lib.cam.ac.uk', apikey: '385784fa-0c04-11e4-b5cc-a3745fee5631' }
  , { id: 3, username: 'genizah', password: 'secret', email: 'genizah@lib.cam.ac.uk', apikey: '436b3396-0c04-11e4-b7ed-0308c2efa56c' }
];

module.exports = users;
