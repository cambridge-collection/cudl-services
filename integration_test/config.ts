export const connectionDetails = {
  host: process.env.TEST_DB_HOST || 'localhost',
  port: Number(process.env.TEST_DB_PORT) || 5432,
  user: process.env.TEST_DB_USER || 'cudlservices',
  password: process.env.TEST_DB_PASSWORD || 'password',
  database: process.env.TEST_DB_DATABASE || 'cudlservices',
};

export const awsEndpointUrl =
  process.env.TEST_AWS_ENDPOINT_URL || 'http://localhost:4566';
