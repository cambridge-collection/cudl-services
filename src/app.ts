import { XSLTExecutor } from '@lib.cam/xslt-nailgun';
import bodyParser from 'body-parser';
import Debugger from 'debug';
import express, { NextFunction, Request, Response } from 'express';
import passport from 'passport';
import path from 'path';
import { CollectionDAO, PostgresCollectionDAO } from './collections';
import { Config, User, Users } from './config';

import { DAOPool, PostgresDatabasePool } from './db';
import {
  CUDLMetadataRepository,
  DefaultCUDLMetadataRepository,
  LegacyDarwinMetadataRepository,
} from './metadata';
import { BaseResource, ExternalResources, using } from './resources';
import * as darwin from './routes/darwin';
import * as membership from './routes/membership';
import * as metadata from './routes/metadata';
import * as similarity from './routes/similarity';
import * as tags from './routes/tags';
import { PostgresTagsDAO, TagsDAO } from './routes/tags-impl';
import * as transcription from './routes/transcription';
import * as translation from './routes/translation';
import { DefaultXTF, XTF } from './xtf';

const cookieParser = require('cookie-parser');
const favicon = require('serve-favicon');
const logger = require('morgan');
const Strategy = require('passport-accesstoken').Strategy;

const debug = Debugger('cudl-services');

// FIXME: move this to server?
//cache directories
// fs.ensureDirSync(config.cacheDir);
// fs.ensureDirSync(config.cacheDir+'/transcriptions');
// fs.ensureDirSync(config.cacheDir+'/translations');

export interface AppOptions {
  users: Users;
  metadataRepository: CUDLMetadataRepository;
  legacyDarwinMetadataRepository: LegacyDarwinMetadataRepository;
  collectionsDAOPool: DAOPool<CollectionDAO>;
  tagsDAOPool: DAOPool<TagsDAO>;
  darwinXtfUrl: string;
  xtf: XTF;
}

export class App extends BaseResource {
  readonly options: AppOptions;
  readonly expressApp: express.Application;
  protected readonly xsltExecutor: XSLTExecutor;

  constructor(options: AppOptions) {
    super();
    this.options = Object.freeze({ ...options });
    this.xsltExecutor = XSLTExecutor.getInstance();
    this.expressApp = this.createExpressApp();
  }

  static fromConfig(config: Config): ExternalResources<App> {
    const dbPool = PostgresDatabasePool.fromConfig(config);

    return new ExternalResources(
      new App({
        metadataRepository: new DefaultCUDLMetadataRepository(config.dataDir),
        legacyDarwinMetadataRepository: new LegacyDarwinMetadataRepository(
          config.legacyDcpDataDir
        ),
        collectionsDAOPool: PostgresCollectionDAO.createPool(dbPool),
        tagsDAOPool: PostgresTagsDAO.createPool(dbPool),
        users: config.users,
        darwinXtfUrl: config.darwinXTF,
        xtf: new DefaultXTF(config),
      }),
      [dbPool]
    );
  }

  private createExpressApp(): express.Application {
    const app = express();

    passport.use(
      new Strategy((token: string, done: (err: any, user: any) => void) => {
        process.nextTick(() => {
          const user = findByApiKey(this.options.users, token);
          return done(null, user || false);
        });
      })
    );

    // view engine setup
    app.set('views', path.resolve(__dirname, '../views'));
    app.set('view engine', 'pug');

    app.use(
      favicon(path.resolve(__dirname, '../public/images/brand/favicon.ico'))
    );
    app.use(logger('dev'));
    app.use(bodyParser.json());
    app.use(bodyParser.urlencoded({ extended: false }));
    app.use(cookieParser());
    app.use(passport.initialize());
    app.use(express.static(path.resolve(__dirname, '../public')));

    // Middleware to redirect trailing slashes to same URL without trailing slash
    app.use((req, res, next) => {
      if (req.url.substr(-1) === '/' && req.url.length > 1) {
        res.redirect(301, req.url.slice(0, -1));
      } else {
        next();
      }
    });

    app.use(
      '/v1/metadata',
      metadata.getRoutes({
        metadataRepository: this.options.metadataRepository,
      })
    );

    app.use(
      '/v1/rdb/membership',
      membership.getRoutes({
        getItemCollections: async (itemID: string) =>
          using(this.options.collectionsDAOPool.getInstance(), dao =>
            dao.getItemCollections(itemID)
          ),
      })
    );

    app.use('/v1/tags', tags.getRoutes({ daoPool: this.options.tagsDAOPool }));

    app.use(
      '/v1/transcription',
      transcription.getRoutes({
        metadataRepository: this.options.metadataRepository,
        legacyDarwinMetadataRepository: this.options
          .legacyDarwinMetadataRepository,
        xsltExecutor: this.xsltExecutor,
      })
    );

    app.use(
      '/v1/translation',
      translation.getRoutes({
        metadataRepository: this.options.metadataRepository,
        xsltExecutor: this.xsltExecutor,
      })
    );

    app.use(
      '/v1/xtf/similarity',
      similarity.getRoutes({
        metadataRepository: this.options.metadataRepository,
        xtf: this.options.xtf,
      })
    );

    app.use(
      '/v1/darwin',
      passport.authenticate('token', { session: false }),
      darwin.getRoutes({ darwinXtfUrl: this.options.darwinXtfUrl })
    );

    // 404 if no route matched
    app.use((req: Request, res: Response, next: NextFunction) => {
      res.status(404).send('Not Found');
    });

    return app;
  }

  async close(): Promise<void> {
    await Promise.all([super.close(), this.xsltExecutor.close()]);
  }
}

function findByApiKey(users: Users, apiKey: string): User | null {
  if (apiKey in users) {
    return users[apiKey];
  }
  return null;
}
