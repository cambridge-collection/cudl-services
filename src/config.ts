export const CONFIG_ENVAR = 'CUDL_SERVICES_CONFIG';

export function loadConfigFromEnvar(): Config {
  const configModulePath = process.env[CONFIG_ENVAR];
  if (!configModulePath) {
    throw new Error(`\
Configuration not found: envar ${CONFIG_ENVAR} be set to the path of the \
config module to load`);
  }

  return require(configModulePath);
}

export interface Config {
  dataDir: string;
  users: Users;
  darwinXTF: string;
  [index: string]: unknown;
}

export interface Users {
  [apiKey: string]: User;
}

export interface User {
  username: string;
  email: string;
}
