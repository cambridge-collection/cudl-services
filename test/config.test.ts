import {Config, isConfig, loadConfigFromModule} from '../src/config';

jest.mock(
  'example-config-module',
  () => {
    const config: Config = {
      createApplication: jest.fn(),
    };
    return config;
  },
  {virtual: true}
);

jest.mock(
  'example-invalid-config-module',
  () => {
    return {};
  },
  {virtual: true}
);

describe('loadConfigFromModule', () => {
  test('returns Config from valid module reference', async () => {
    const expectedConfig = require('example-config-module') as Config;

    await expect(loadConfigFromModule('example-config-module')).resolves.toBe(
      expectedConfig
    );
  });

  test('raises error when module does not exist', async () => {
    await expect(
      loadConfigFromModule('does/not/exist/sdfsdfa')
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  test('raises error when module does not export a Config object', async () => {
    await expect(
      loadConfigFromModule('does/not/exist/sdfsdfa')
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});

test.each([
  [undefined, false],
  [42, false],
  [{}, false],
  [{createApplication: undefined}, false],
  [{createApplication: 42}, false],
  [{createApplication: () => {}}, true],
])('isConfig(%j) = %j', (value, expected) => {
  expect(isConfig(value)).toBe(expected);
});
