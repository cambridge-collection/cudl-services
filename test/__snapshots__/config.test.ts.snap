// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadConfigFromModule raises error when module does not exist 1`] = `"Failed to load config from module \\"does/not/exist/sdfsdfa\\": Error: Cannot find module 'does/not/exist/sdfsdfa' from 'src/config.ts'"`;

exports[`loadConfigFromModule raises error when module does not export a Config object 1`] = `"Failed to load config from module \\"does/not/exist/sdfsdfa\\": Error: Cannot find module 'does/not/exist/sdfsdfa' from 'src/config.ts'"`;
